### ISSUE TITLE
False negative on raising-bad-type

### ISSUE THREAD

**Issue by eirnym** [PyCQA/pylint#2529]
raising-bad-type doesn't check if type was checked already with `isinstance()`.
trivia: `None` is not an instance of `Exception`. 

### Steps to reproduce:

```python
import random

def process_exception():
    e = None
    if random.randint(0, 1):
        e = Exception()

    if isinstance(e, Exception):
        raise e
```
### Current behavior
```
test.py:9:8: E0702: Raising NoneType while only classes or instances are allowed (raising-bad-type)
```
### Expected behavior
No errors as type is checked

### pylint --version output
```
pylint 2.1.1
astroid 2.0.4
Python 3.7.0 (default, Jun 28 2018, 06:05:58) 
[Clang 9.0.0 (clang-900.0.39.2)]
```

**Comment by eirnym** [comment]
Nevertheless, some checks are aware about control flow

**Comment by PCManticore** [comment]
True, but it's not feasible to add this for all the checks we have, as it increases the maintenance burden with what I'd consider them to be hacks. We'll need to solve this problem at the source, in `astroid`.
